File: strands-ts/src/sandbox/docker.ts
/**
 * Docker sandbox — executes commands in a Docker container via `docker exec`.
 */

import type { ExecuteOptions } from './base.js'
import { PosixShellSandbox, validateEnvKeys } from './posix-shell.js'
import { streamProcess } from './stream-process.js'
import type { ExecutionResult, StreamChunk } from './types.js'
import type { Tool } from '../tools/tool.js'
import { makeFileEditor, DEFAULT_FILE_EDITOR_DESCRIPTION } from '../vended-tools/file-editor/index.js'
import { makeBash, SANDBOX_BASH_DESCRIPTION } from '../vended-tools/bash/index.js'

/**
 * Options for constructing a {@link DockerSandbox}.
 */
export interface DockerSandboxOptions {
  /** ID or name of a running Docker container. */
  container: string
  /**
   * Working directory for executed commands. If omitted, no `-w` flag is set and
   * commands run in the container's configured working directory.
   *
   * Set this to run elsewhere; the path must exist and be writable by the effective {@link user}.
   */
  workingDir?: string
  /**
   * User to run commands as, in `"uid"`, `"uid:gid"`, or `"name"` form. If omitted,
   * no `--user` flag is set and commands run as the container's configured user.
   *
   * Set this to override the container's identity, e.g. `"root"` or `"1000:1000"`.
   */
  user?: string
}

/** Execute commands in a Docker container via `docker exec`. */
export class DockerSandbox extends PosixShellSandbox {
  readonly container: string
  readonly workingDir: string | undefined
  private readonly _user: string | undefined

  constructor(options: DockerSandboxOptions) {
    super()
    this.container = options.container
    this.workingDir = options.workingDir
    this._user = options.user
  }

  async *executeStreaming(
    command: string,
    options?: ExecuteOptions
  ): AsyncGenerator<StreamChunk | ExecutionResult, void, undefined> {
    const args = ['exec']

    // Unset user/cwd defer to the container's own configuration.
    if (this._user !== undefined) args.push('--user', this._user)

    const cwd = options?.cwd ?? this.workingDir
    if (cwd !== undefined) args.push('-w', cwd)

    if (options?.env) {
      validateEnvKeys(options.env)
      for (const [key, val] of Object.entries(options.env)) {
        // Values are passed as process argv (not through a shell), so no escaping is
        // needed -- Docker stores them verbatim. This is why values aren't shell-quoted
        // here, unlike the SSH backend which builds a shell command string.
        args.push('-e', `${key}=${val}`)
      }
    }

    // docker exec requires the container and command after all flags. Furthermore, '--'
    // terminates flag parsing so the container is always treated as a positional argument.
    // Without this, a name like '--privileged' would be parsed as a flag, overriding the
    // exec options above.
    args.push('--', this.container, 'sh', '-c', command)

    yield* streamProcess('docker', args, {
      timeout: options?.timeout,
      signal: options?.signal,
      enoentMessage: 'docker is not installed or not on PATH',
    })
  }

  override getTools(): Tool[] {
    const cwd = this.workingDir ? ` Working directory: ${this.workingDir}.` : ''
    return [
      makeFileEditor(this, {
        description: `${DEFAULT_FILE_EDITOR_DESCRIPTION} Files are in Docker container "${this.container}".`,
      }),
      makeBash(this, {
        description: `${SANDBOX_BASH_DESCRIPTION} Runs in Docker container "${this.container}".${cwd}`,
      }),
    ]
  }
}

File: strands-py/src/strands/sandbox/docker.py
"""Docker sandbox -- executes commands in a Docker container via ``docker exec``.

Mirrors ``strands-ts/src/sandbox/docker.ts``.
"""

from collections.abc import AsyncGenerator
from typing import Any

from .shell import PosixShellSandbox, validate_env_keys
from .stream_process import stream_process
from .types import ExecutionResult, StreamChunk


class DockerSandbox(PosixShellSandbox):
    """Execute commands in a Docker container via ``docker exec``.

    A thin :class:`PosixShellSandbox` backend: file and code operations are
    inherited (run as shell commands), and only :meth:`execute_streaming` is
    implemented, building the ``docker exec`` argv.
    """

    def __init__(self, container: str, *, working_dir: str | None = None, user: str | None = None) -> None:
        """Initialize the Docker sandbox.

        Args:
            container: ID or name of a running Docker container.
            working_dir: Working directory for executed commands. If ``None``, no
                ``-w`` flag is set and commands run in the container's configured
                working directory. The path must exist and be writable by the
                effective ``user``.
            user: User to run commands as, in ``"uid"``, ``"uid:gid"``, or
                ``"name"`` form. If ``None``, no ``--user`` flag is set and
                commands run as the container's configured user.
        """
        self.container = container
        self.working_dir = working_dir
        self._user = user

    async def execute_streaming(
        self,
        command: str,
        *,
        timeout: float | None = None,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        **kwargs: Any,
    ) -> AsyncGenerator[StreamChunk | ExecutionResult, None]:
        """Execute a command in the container, streaming output.

        Args:
            command: The shell command to execute.
            timeout: Maximum execution time in seconds. ``None`` means no timeout.
            cwd: Working directory for this command, overriding ``working_dir``.
            env: Environment variables to set, passed as ``docker exec -e`` flags.
            **kwargs: Additional keyword arguments for forward compatibility.

        Yields:
            :class:`StreamChunk` objects for output, then a final
            :class:`ExecutionResult`.

        Raises:
            ValueError: If an environment variable name is invalid.
            TimeoutError: If execution exceeds ``timeout`` seconds.
        """
        args = ["exec"]

        # An unset user/cwd defers to the container's own configuration.
        if self._user is not None:
            args += ["--user", self._user]

        effective_cwd = cwd if cwd is not None else self.working_dir
        if effective_cwd is not None:
            args += ["-w", effective_cwd]

        if env:
            validate_env_keys(env)
            for key, value in env.items():
                # Values are passed as process argv (not through a shell), so no escaping
                # is needed -- Docker stores them verbatim. This is why values are not
                # shell-quoted here, unlike the SSH backend which builds a shell string.
                args += ["-e", f"{key}={value}"]

        # docker exec requires the container and command after all flags. '--' terminates
        # flag parsing so the container is always treated as a positional argument; without
        # it, a name like '--privileged' would be parsed as a flag and override the options
        # above.
        args += ["--", self.container, "sh", "-c", command]

        async for chunk in stream_process(
            "docker", args, timeout=timeout, enoent_message="docker is not installed or not on PATH"
        ):
            yield chunk
